flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -150,6 +164,19 @@ class ReachCollector {
     }
   }
 
+  /// Is \p op in the region of the array expression.
+  inline bool inArrayExprRegion(mlir::Operation *op) {
+    return domInfo.dominates(exprTop.getOperation(), op) &&
+           postDomInfo.postDominates(exprBtm.getOperation(), op);
+  }
+
+  /// Is \p op in the elemental computation body of the array expression.
+  inline bool inArrayExprBody(mlir::Operation *op) {
+    return domInfo.dominates(exprTop.getOperation(), op) &&
+           !domInfo.dominates(op, exprBtm.getOperation()) &&
```

Comment:
Wouldn't `inArrayExprBody` also return true if OP is computed inside an if-op in the array expression lowering prelude (after the top array load, but before the loops) ? e.g. like `OP` in the pseudo fir below:

```
%top = fir.array_load %memref
...
%something = fir.if {
  // not the array expr body yet (e.g could be some scalar index calculation in "x + y(merge(foo(), bar(), zorg()), :)"
  %x = OP
  fir.result %OP
} else {
  ...
}
%array_operand = fir.array_load .... %something

%loop = fir.do_loop .... {
   // array expr body
}
fir.array_merge_store %loop, %top .....
```